File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxPathSum = Integer.MIN_VALUE;
    public int helper(TreeNode root) {
        if(root == null)
            return 0;
        int left = helper(root.left);
        int right = helper(root.right);
        
        int ans = Math.max(left, right) + root.val;
        ans = Math.max(ans, root.val);

        maxPathSum = Math.max(maxPathSum, left + right + root.val);
        maxPathSum = Math.max(maxPathSum, ans);
        return ans;
    }
    public int maxPathSum(TreeNode root) {
        helper(root);
        return maxPathSum;
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxSum = float('-inf')
        def dfs(root):
            if root is None:
                return 0
            l = dfs(root.left)
            r = dfs(root.right)
            nodeSum = max(max(l, r) + root.val, root.val)
            self.maxSum = max(self.maxSum, l + r + root.val)
            self.maxSum = max(self.maxSum, nodeSum)
            return nodeSum
        dfs(root)
        return self.maxSum